linera-service/src/chain_listener.rs

```diff
@@ -37,56 +61,137 @@ where
     S: Store + Clone + Send + Sync + 'static,
     ViewError: From<S::ContextError>,
 {
-    /// Creates a new chain listener given a client chain.
-    pub fn new(config: ChainListenerConfig, client: Arc<Mutex<ChainClient<P, S>>>) -> Self {
-        Self { config, client }
+    /// Creates a new chain listener given client chains.
+    pub(crate) fn new(config: ChainListenerConfig, clients: ClientMap<P, S>) -> Self {
+        Self { config, clients }
     }
 
     /// Runs the chain listener.
-    pub async fn run<C, F>(self, mut context: C, wallet_updater: F) -> Result<(), anyhow::Error>
+    pub async fn run<C, F>(
+        self,
+        mut context: C,
+        wallet_updater: F,
```

Comment:
Shouldn't `update_wallet` be a method of the trait `ClientContext` since you created a trait :) ?

Reply:
Done.